source/materials/OpticalMaterialProperties.cc

```diff
@@ -21,16 +23,19 @@ G4MaterialPropertiesTable* OpticalMaterialProperties::Vacuum()
 {
   G4MaterialPropertiesTable* mpt = new G4MaterialPropertiesTable();
 
-  const G4int entries = 2;
-  G4double energy[entries] = {optPhotMinE_, optPhotMaxE_};
+  std::vector<G4double> energy = {optPhotMinE_, optPhotMaxE_};
 
   // REFRACTIVE INDEX
-  G4double rindex[entries] = {1., 1.};
-  mpt->AddProperty("RINDEX", energy, rindex, entries);
+  std::vector<G4double> ri_index = {1., 1.};
+
+  assert(energy.size() == ri_index.size());
+  mpt->AddProperty("RINDEX", energy.data(), ri_index.data(), energy.size());
```

Comment:
If the check is included in the `AddProperty` function, I don't see the point of writing a function now to optimize a piece of code that will become obsolete soon.

Reply:
You wrote out this pattern by hand *54 times*!

After you wrote it out 3 or 4 times (ideally, after the second time) you should have said to yourself, "I *refuse* to type this out by hand again!" and you should have written the abstraction.

As it stands, there is a small but greater than zero chance that some of your asserts check different values than are passed to `AddProperty`. (*I* certainly couldn't be bothered to check them all! Even if I did, it's so repetitively tedious, that I'd be very likely to miss any mistakes.) By using the abstraction, you reduce the probability of such a mistake to strictly zero. As these are run-time rather than compile-time checks, they are more difficult to spot, and have consequences that are nastier to deal with, so it is especially important to avoid the mistake.

Using the abstraction will shorten the code by 54 lines, while writing the abstraction will cost 3 non-brace-only lines.

You are expecting this code to be updated for 10.7. Think of the poor soul who will have to do it, and will be faced with 54 instances of 

``` c++
assert(energy.size() == ri_index.size());
mpt->AddProperty("RINDEX", energy.data(), ri_index.data(), energy.size());
```

which will need to be changed to 

``` c++
mpt->AddProperty("RINDEX", energy, ri_index);
```

Why not be nice to that person (probably you) and leave 54 instances of 

``` c++
AddProperty(mpt, "RINDEX", energy, ri_index);
```

Better still, with the abstraction in place, when the time comes to switch to 10.7, you can change the *definition* of your abstraction (removing the assert, and switching to the `std::vector`-using overload) just once, and leave its 54 uses untouched, and you'll have all the benefits of the 10.7 version at the cost of changing 2 lines. As it stands, switching to the 10.7 version will require changing 108 lines.

You may argue that there is no point in changing 108 lines *now* to save changing 108 lines later. There are two answers to this:

1. It is now fresh in your mind and you know exactly what is happening. Leave it in, and you will have to re-establish the mental context, months down the line; or, worse still, someone *else* will have to acquire the mental context. Doing it now is cheaper than doing it in the future.

2. You shouldn't have written out 54 repetitions of something that will require 54 fixes, in the first place. It's our old friend copy-paste: just don't do it!

In summary, it's simply a very basic example of good programming practice that you should follow, because it is good for you and anyone else whose life is touched by the code. It is a *habit* that you should develop, because it will result in you writing more robust code, more quickly, that is more pleasant to work with both for you and for your collaborators.

Put another way, *I* don't see the point of writing bad code, when it's so easy to write better code.

Important point: you should have refused to write the assert out 54 times, in the first place!

Important point to *everyone*: please *refuse* to write out the same tedious stuff, over and over and over again. Think to yourselves: what interface would make my life easier? ... and then *create* that interface.

**Copy paste? JUST SAY NO!**

Reply:
> You may argue that there is no point in changing 108 lines now to save changing 108 lines later.

That was exactly my point :-).

I agree with you that I shouldn't have made the effort of writing 54 times the same pattern. In fact, I shouldn't have changed the code at all at this point and leave the implementation of vectors instead of C-arrays for the update to Geant4.10.7, but your point of being educational convinced me.

I'm not sure it is better to be left with 

```
AddProperty(mpt, "RINDEX", energy, ri_index);
```

once we pass to Geant4.10.7, rather than the current

```
mpt->AddProperty("RINDEX", energy, ri_index);
```

which seems more straightforward to me.

Anyway, I would finish the review of this PR on the physics content (@carmenromo is assigned) and I would leave for another PR any implementation you would like to do.

Reply:
Fine. I'll decouple from this PR.